addons/stock/models/stock_picking.py

```diff
@@ -830,6 +841,158 @@ def _create_backorder(self, backorder_moves=[]):
+        """ Generic method to log activity. It either log on uppermost
+        ongoing documents or following documents. From a dict
+        containing multiple records and their changes, get impacted
+        objects based on a specific field and group them by common
+        document and responsible. Then log an activity for each pair
+        document, responsible with a note containing only related data.
+        example: We want to notify all the delivery picking if a
+        receipt picking is modified. We will pass the modified moves
+        and their quantity.
+        {'move_1': (4, 5), 'move_2': (3, 4), 'move_3': (2, 2)}.
+        move_1 and move_2 are for picking 1 and move_3 on picking 2.
+        receipt moves and delivery moves are linked with the
+        move_dest_ids field thus we use it has 'field_to_group'.
+        Thus this fuction will log activity on picking 1 with
+        information about move 1, 2 and another activity on picking 2
+        with information about move 3.
+
+        :param orig_obj_changes dict: contain a record as key and the
+        change on this record as value.
+        eg: {'move_id': (new product_uom_qty, old product_uom_qty)}
+        :param field_to_group string: It has to be a field of the
+        records that are register in the key of 'orig_obj_changes'
+        eg: 'move_dest_ids' if we use move as record (previous example)
+        :param note_method method: a static function that will generate
+        the html note to log on the activity. The note_method should
+        use the args:
+            - orig_obj_changes
+            - a recordset of a one2many field on which the note is
+            logged. (eg: a list of moves on a picking.)
+            - a dict with link between the orig_obj_changes and the
+            recordset passed as second argument. (eg: If we want to
+            notify the missing quantity from a move to all moves dest
+            we wil pass the moves dest as the recodset but we don't
+            know what are the quantity that changed on the original
+            move. that's the purpose of this dict that will return
+            the original move due to the destination move and we will
+            be able to use the orig_obj_changes in order to find the
+            modified quantity.)
+        the note_method should return a string with an html format
+        :param stream string:
+            - 'UP' if we want to log on the upper most ongoing
+            documents.
+            - 'DOWN' if we want to log on following documents.
+        :param sorted_method method, groupby_method: Only need when
+        stream is 'DOWN', it should sort/group by (object on which the
+        activity is log, the responsible for this object)
+        """
+        orig_obj_child_object_rel = {co: ooc for ooc in orig_obj_changes for co in ooc[field_to_group]}
```

Comment:
maybe use .keys() to make it clear that the first part loops on the keys of the dict